### PR TITLE
Override feature to allow users to manually add/remove wells from projects

### PR DESCRIPTION
**Pull request by apd-ypuranik**
## Fixes/Addresses/Summary/Motivation:

Implements an "override" feature that allows users to manually update optimization results

## Changes proposed in this PR:
- An example of tests for ipywidgets using playwright and solara
- Working demo with override feature partially implemented

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my
contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

### CODE REVIEW

File: primo/utils/override_utils.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+# User-defined libs
+from primo.data_parser.data_model import OptInputs
+from primo.utils.raise_exception import raise_exception
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Recalculate:
+    """
+    Class for assess whether whether the overridden P&A projects adhere to the constraints
```
Comment (Jaffe11) [inline primo/utils/override_utils.py:31]: assess -> assessing and remove the second whether

File: primo/utils/override_utils.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    defined in the optimization problem
+
+    Parameters
+    ----------
+
+    selected_wells : pd.DataFrame
+        A DataFrame containing wells selected based on solving the optimization problem
+        and/or by manual selections and overrides
+
+    wells_added : list
```
Comment (Jaffe11) [inline primo/utils/override_utils.py:41]: list -> List[str]

File: primo/utils/override_utils.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        A list of wells that the user wishes to add to the P&A projects
+
+    wells_removed : list
```
Comment (Jaffe11) [inline primo/utils/override_utils.py:44]: list -> List[str]

File: primo/utils/override_utils.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        leave room for budget, the method returns a list of candidates that can
+        be added to the plugging projects without violating any constraints
+
+        Parameters:
+        ----------
+        None
+
+        Returns:
+        --------
+        List of Well API numbers that can be added in the project
```
Comment (Jaffe11) [inline primo/utils/override_utils.py:265]: Well -> well

File: primo/utils/override_utils.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        additions = []
+        if self.assess_feasibility() is False:
+            # The
```
Comment (Jaffe11) [inline primo/utils/override_utils.py:269]: This comment either needs to be finished or removed.

**Review by Jaffe11 — CHANGES_REQUESTED** [review]

**Review by strahlw — CHANGES_REQUESTED** [review]

File: primo/demo/PRIMO - Example_1.ipynb [hunk 6]
```diff
@@ -5242,86 +5667,748 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a389689e",
+   "metadata": {},
+   "source": [
+    "##### Well Suggestions for backfill\n",
+    "\n",
+    "If as a result of manual overrides there is possibility of adding new wells to the plugging projects, PRIMO can make suggestions for new wells to be included!"
```
Comment (Jaffe11) [inline primo/demo/PRIMO - Example_1.ipynb:5782]: "is a possibility"

File: primo/demo/PRIMO - Example_1.ipynb [hunk 7]
```diff
@@ -5242,86 +5667,748 @@
...
+   "source": [
+    "violation_oil = Recalculate(selected_oil, well_add_list_oil, well_remove_list_oil,oil_gdf,opt_inputs_oil)\n",
+    "\n",
+    "violate_cost_oil=violation_oil.assess_budget()\n",
+    "violate_operator_oil=violation_oil.assess_owner_well_count()\n",
+    "violate_distance_oil=violation_oil.assess_distances()\n",
+    "\n",
+    "if violate_cost_oil > 0:\n",
+    "    display(Markdown(f\"\"\" After the modification, the total budget is over the limit by ${int(violate_cost_oil)}.\n",
+    "                     Please consider to modify wells you have selected use the widget above or re-run the optimization problem.\"\"\"))\n",
```
Comment (Jaffe11) [inline primo/demo/PRIMO - Example_1.ipynb:6153]: "Please consider modifying the wells you have selected by using either the widget above, or by re-running the optimization problem"

File: primo/demo/PRIMO - Example_1.ipynb [hunk 8]
```diff
@@ -5242,86 +5667,748 @@
...
+    "\n",
+    "if bool(violate_operator_oil) is not False:\n",
+    "    display(Markdown(f\"\"\" After the modification, the following owners have more than {max_wells_per_owner} well(s) \n",
+    "                     being selected. Please consider to modify wells you have selected use the widget above\n",
```
Comment (Jaffe11) [inline primo/demo/PRIMO - Example_1.ipynb:6157]: Same correction here: "Please consider modifying the wells you have selected by using either the widget above, or by re-running the optimization problem"

File: primo/demo/PRIMO - Example_1.ipynb [hunk 9]
```diff
@@ -5242,86 +5667,748 @@
...
+    "                     or re-run the optimization problem.\"\"\"))\n",
+    "    violate_operator_oil_df = pd.DataFrame.from_dict(violate_operator_oil,orient='index')\n",
+    "    violate_operator_oil_df = violate_operator_oil_df.reset_index()\n",
+    "    violate_operator_oil_df = violate_operator_oil_df.rename(columns={'index': 'Owner', 0: 'Number of wells', 1: 'Wells'})\n",
+    "    display(violate_operator_oil_df)\n",
+    "\n",
+    "\n",
+    "if bool(violate_distance_oil) is not False:\n",
+    "    display(Markdown(f\"\"\" After the modification, the following projects have wells are far away from each others. \n",
+    "                     Please consider to modify wells you have selected use the widget above\n",
```
Comment (Jaffe11) [inline primo/demo/PRIMO - Example_1.ipynb:6167]: Same correction here: "Please consider modifying the wells you have selected by using either the widget above, or by re-running the optimization problem"

**Review by Jaffe11 — CHANGES_REQUESTED** [review]

File: primo/utils/override_utils.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+        ]
+
+        # Remove candidates in removal list
+        candidates = candidates[~candidates["API Well Number"].isin(self.wells_removed)]
+        # Stick to those projects that already exist
+        existing_projects = set(self._plugged_list["Project"])
+        candidates = candidates[candidates["Project"].isin(existing_projects)]
+
+        for well_id in candidates["API Well Number"]:
+            self._add_well(well_id)
```
Comment (ruonanli61) [inline primo/utils/override_utils.py:287]: We might need to clarify the function of the backfill feature. Do we also provide backfill candidates to users when some of the wells added to the suggested projects violate the constraint?

**Review by ruonanli61 — CHANGES_REQUESTED** [review]

File: primo/utils/override_utils.py (hunk 1 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:31]: Fixed

File: primo/utils/override_utils.py (hunk 2 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:41]: updated

File: primo/utils/override_utils.py (hunk 3 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:44]: updated

File: primo/utils/override_utils.py (hunk 4 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:265]: Fixed

File: primo/utils/override_utils.py (hunk 5 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:269]: ha thanks :-) Fixed!

File: primo/demo/PRIMO - Example_1.ipynb [hunk 11]
```diff
@@ -5242,86 +5667,748 @@
...
+    "                     or re-run the optimization problem.\"\"\"))\n",
+    "    violate_distance_gas_df = pd.DataFrame.from_dict(violate_distance_gas,orient='index')\n",
+    "    violate_distance_gas_df = violate_distance_gas_df.reset_index(drop=True)\n",
+    "    violate_distance_gas_df = violate_distance_gas_df.rename(columns={0: 'Project', 1: 'Well 1', 2: 'Well 2', 3: \"Distance between Well 1 and 2 [Miles]\"})\n",
+    "    display(violate_distance_gas_df)\n",
+    "\n",
+    "new_df_or_gas,cols_to_display,grouped_or_gas,total_cost_or_gas=generate_summary(violation_gas._plugged_list,max_val_col,min_val_col,efficiency_weights)\n",
+    "display(Markdown(\"\"\" A summary of the results based on the override well list selected\"\"\"))\n",
+    "if violation_gas.assess_feasibility():\n",
+    "    display(Markdown(\"\"\" Project Status: FEASIBLE\"\"\"))\n",
```
[the remark this answers is not in the document]
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:5768]: Agreed. I expect a whole range of changes to how we present override features: We can address this in a future PR if you agree

File: primo/demo/PRIMO - Example_1.ipynb [hunk 12]
```diff
@@ -5242,86 +5667,748 @@
...
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "if violation_gas.assess_feasibility() is False:\n",
+    "    display(\n",
+    "        Markdown(\n",
+    "            f\"\"\" No backfill candidates can be generated due to infeasibility. Please re-optimize\"\"\"\n",
```
[the remark this answers is not in the document]
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:5808]: Same as above: expect to address this in a future PR

File: primo/demo/PRIMO - Example_1.ipynb [hunk 13]
```diff
@@ -5169,57 +5326,325 @@
...
+   "source": [
+    "#### Gas wells"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0e07bcb0",
+   "metadata": {},
+   "source": [
+    "##### Select Wells Would Like to be Included in or Excluded from the Recommended P&A Projects \n",
```
[the remark this answers is not in the document]
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:5351]: fixed!

File: primo/demo/PRIMO - Example_1.ipynb [hunk 14]
```diff
@@ -5242,86 +5667,748 @@
...
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "if violation_oil.assess_feasibility() is False:\n",
+    "    display(\n",
+    "        Markdown(\n",
+    "            f\"\"\" No backfill candidates can be generated due to infeasibility. Please re-optimize\"\"\"\n",
```
[the remark this answers is not in the document]
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:6219]: same response about technical language and new PR :-)

File: primo/demo/PRIMO - Example_1.ipynb (hunk 6 above)
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:5782]: fixed!

File: primo/demo/PRIMO - Example_1.ipynb (hunk 7 above)
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:6153]: Fixed!

File: primo/demo/PRIMO - Example_1.ipynb (hunk 8 above)
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:6157]: Fixed

File: primo/demo/PRIMO - Example_1.ipynb (hunk 9 above)
Comment (apd-ypuranik) [reply primo/demo/PRIMO - Example_1.ipynb:6167]: Fixed

File: README.md [hunk 15]
```diff
@@ -27,36 +27,42 @@ Our complete documentation is available on [readthedocs](https://primo.readthedo
 
 While not required, we encourage the installation of [Anaconda](https://www.anaconda.com/products/individual#Downloads) or [Miniconda](https://docs.conda.io/en/latest/miniconda.html) and using the `conda` command to create a separate python environment in which to install the PRIMO Toolkit.
 
-Use conda to create a new "primo-env" (could be any name you like) environment then activate that environment:
+Regular uses can use conda to create a new "primo" environment.
```
[the remark this answers is not in the document]
Comment (apd-ypuranik) [reply README.md:30]: Fixed!

File: primo/utils/override_utils.py (hunk 10 above)
Comment (apd-ypuranik) [reply primo/utils/override_utils.py:287]: At the moment, we don't! This is clarified in the check in line 268 (and the added comment).

File: primo/demo/PRIMO - Example_1.ipynb (hunk 11 above)
[the remark this answers is not in the document]
Comment (strahlw) [reply primo/demo/PRIMO - Example_1.ipynb:5768]: That's fine with me.

File: primo/demo/PRIMO - Example_1.ipynb (hunk 12 above)
[the remark this answers is not in the document]
Comment (strahlw) [reply primo/demo/PRIMO - Example_1.ipynb:5808]: Agreed.

File: primo/demo/PRIMO - Example_1.ipynb (hunk 14 above)
[the remark this answers is not in the document]
Comment (strahlw) [reply primo/demo/PRIMO - Example_1.ipynb:6219]: agreed.

**Review by strahlw — APPROVED** [review]

**Review by DevKakkad32 — APPROVED** [review]

**Review by Jaffe11 — APPROVED** [review]

**Pull request merged**